controller/rest/longpoll.go

```diff
@@ -52,54 +58,77 @@ func (j *longpollOnceJob) Poll() (interface{}, time.Duration) {
 	}
 }
 
+type PollWorkItem struct {
+	key  interface{}
+	args interface{}
+	task longpollOnceTask
+}
+
 type longpollOnceMgr struct {
-	mux     sync.Mutex
-	max     int
-	timeout time.Duration
-	linger  time.Duration
-	jobs    map[interface{}]*longpollOnceJob
+	mux       sync.RWMutex
+	poolSize  int
+	timeout   time.Duration
+	linger    time.Duration
+	jobs      map[interface{}]*longpollOnceJob // use as a queue
+	workQueue chan PollWorkItem
 }
 
-func NewLongPollOnceMgr(timeout, linger time.Duration, max int) *longpollOnceMgr {
-	return &longpollOnceMgr{
-		max:     max,
-		timeout: timeout,
-		linger:  linger,
-		jobs:    make(map[interface{}]*longpollOnceJob),
+func (m *longpollOnceMgr) pollWorker() {
+	for workItem := range m.workQueue {
+		m.mux.RLock()
+		job, exists := m.jobs[workItem.key]
+		m.mux.RUnlock()
+
+		if exists {
+			job.dataCh <- workItem.task.Run(workItem.args)
+			// In case polling returned, wait a while before discarding the record
+			select {
+			case <-time.After(m.linger):
+			case <-job.polled:
+			}
+			m.mux.Lock()
+			delete(m.jobs, workItem.key)
+			m.mux.Unlock()
+		} else {
+			log.WithFields(log.Fields{
+				"key": workItem.key,
+			}).Error("Job not found in manager")
+		}
+	}
+}
+
+func NewLongPollOnceMgr(timeout, linger time.Duration, poolSize int) *longpollOnceMgr {
+	mgr := &longpollOnceMgr{
+		poolSize:  poolSize,
+		timeout:   timeout,
+		linger:    linger,
+		jobs:      make(map[interface{}]*longpollOnceJob),
+		workQueue: make(chan PollWorkItem, workQueueBufferSize),
+	}
+
+	// Run a work pool to maintain the speed for process the jobs in jobs Map
+	for i := 0; i < poolSize; i++ {
+		go mgr.pollWorker()
 	}
+
+	return mgr
 }
 
 func (m *longpollOnceMgr) GetJobCount() int {
+	m.mux.RLock()
+	defer m.mux.RUnlock()
 	return len(m.jobs)
 }
 
-func (m *longpollOnceMgr) NewJob(key interface{}, task longpollOnceTask, arg interface{}) (*longpollOnceJob, error) {
+func (m *longpollOnceMgr) NewJob(key interface{}, task longpollOnceTask, args interface{}) (*longpollOnceJob, error) {
 	m.mux.Lock()
 	defer m.mux.Unlock()
 
 	job, ok := m.jobs[key]
 	if !ok {
-		if m.max > 0 && len(m.jobs) >= m.max {
-			return nil, errTooManyJobs
-		}
-
 		job = newLongPollOnceJob(m.timeout)
 		m.jobs[key] = job
-
-		go func() {
-			job.dataCh <- task.Run(arg)
-
-			// In case polling returned, wait a while before discarding the record
-			select {
-			case <-time.After(m.linger):
-			case <-job.polled:
-			}
-
-			m.mux.Lock()
-			delete(m.jobs, key)
-			m.mux.Unlock()
-		}()
-
+		m.workQueue <- PollWorkItem{key: key, args: args, task: task}
```

Comment:
Sorry if the indent is not correct:
```suggestion
                    select {
                        case m.workQueue <- PollWorkItem{key: key, args: args, task: task}
                        default:
                            log.Error("workQueue is full")
                    }
```